safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,170 @@
+
+// Nonce retrieves the nonce from the given context. If there is no nonce stored
+// in the context, a new nonce is generated and placed in the context.
+func Nonce(ctx context.Context) (string, context.Context) {
+	v := ctx.Value(ctxKey{})
+	if v == nil {
+		v = generateNonce()
+		ctx = context.WithValue(ctx, ctxKey{}, v)
+	}
+	return v.(string), ctx
+}
+
+// StrictCSPBuilder can be used to build a strict, nonce-based CSP.
+// See https://csp.withgoogle.com/docs/strict-csp.html for more info.
+//
+// If BaseURI is an empty string the base-uri directive will be set to 'none'.
+type StrictCSPBuilder struct {
+	ReportOnly    bool
+	StrictDynamic bool
+	UnsafeEval    bool
+	BaseURI       string
+	ReportURI     string
+}
+
+// Build creates a Policy based on the specified options.
+func (s StrictCSPBuilder) Build() Policy {
+	return Policy{
+		reportOnly: s.ReportOnly,
+		serialize: func(ctx context.Context) (string, context.Context) {
+			var b strings.Builder
+
+			b.WriteString("object-src 'none'; script-src 'unsafe-inline' https: http: 'nonce-")
+			n, ctx := Nonce(ctx)
+			b.WriteString(n)
+			b.WriteString("'")
+
+			if s.StrictDynamic {
+				b.WriteString(" 'strict-dynamic'")
+			}
+			if s.UnsafeEval {
+				b.WriteString(" 'unsafe-eval'")
+			}
+
+			b.WriteString("; base-uri ")
+			if s.BaseURI == "" {
+				b.WriteString("'none'")
+			} else {
+				b.WriteString(s.BaseURI)
+			}
+
+			if s.ReportURI != "" {
+				b.WriteString("; report-uri ")
+				b.WriteString(s.ReportURI)
+			}
+
+			return b.String(), ctx
+		},
+	}
+}
+
+// NewFramingCSP creates a new CSP policy with frame-ancestors set to 'self'.
+//
+// TODO: allow relaxation on specific endpoints according to #77.
+func NewFramingCSP(reportOnly bool) Policy {
+	return Policy{
+		reportOnly: reportOnly,
+		serialize: func(ctx context.Context) (string, context.Context) {
+			return "frame-ancestors 'self'", ctx
+		},
+	}
+}
+
+// Interceptor intercepts requests and applies CSP policies.
+type Interceptor struct {
+	Policies []Policy
+}
+
+// NewInterceptor creates an interceptor from the provided policies.
+func NewInterceptor(p ...Policy) Interceptor {
+	return Interceptor{Policies: p}
+}
+
+// Default creates a new CSP interceptor with a strict nonce-based policy and a
+// framing policy, both in enforcement mode.
+func Default(reportURI string) Interceptor {
+	return NewInterceptor(StrictCSPBuilder{ReportURI: reportURI}.Build(), NewFramingCSP(false))
+}
+
+// Before claims and sets the Content-Security-Policy header and the
+// Content-Security-Policy-Report-Only header.
+func (it Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	h := w.Header()
+	setCSP, err := h.Claim("Content-Security-Policy")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+
+	setCSPReportOnly, err := h.Claim("Content-Security-Policy-Report-Only")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
```

Comment:
Create variables when they're needed, no need to make them here if you are not using them until after the loop.